test/integration/buffer_accounting_integration_test.cc

```diff
@@ -0,0 +1,281 @@
+#include "test/mocks/http/mocks.h"
+
+#include "fake_upstream.h"
+#include "gtest/gtest.h"
+#include "http_integration.h"
+#include "integration_stream_decoder.h"
+#include "socket_interface_swap.h"
+
+namespace Envoy {
+namespace {
+std::string ipVersionAndBufferAccountingTestParamsToString(
+    const ::testing::TestParamInfo<std::tuple<Network::Address::IpVersion, bool>>& params) {
+  return fmt::format(
+      "{}_{}",
+      TestUtility::ipTestParamsToString(::testing::TestParamInfo<Network::Address::IpVersion>(
+          std::get<0>(params.param), params.index)),
+      std::get<1>(params.param) ? "with_per_stream_buffer_accounting"
+                                : "without_per_stream_buffer_accounting");
+}
+} // namespace
+
+class HttpBufferWatermarksTest
+    : public SocketInterfaceSwap,
+      public testing::TestWithParam<std::tuple<Network::Address::IpVersion, bool>>,
+      public HttpIntegrationTest {
+public:
+  struct BufferParams {
+    const uint32_t connection_watermark;
+    const uint32_t downstream_h2_stream_window;
+    const uint32_t downstream_h2_conn_window;
+    const uint32_t upstream_h2_stream_window;
+    const uint32_t upstream_h2_conn_window;
+  };
+
+  // Configures the buffers with the given parameters.
+  void initializeWithBufferConfig(const BufferParams& buffer_params, uint32_t num_responses) {
+    config_helper_.setBufferLimits(buffer_params.connection_watermark,
+                                   buffer_params.connection_watermark);
+
+    config_helper_.addConfigModifier(
+        [&](envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+                hcm) -> void {
+          auto* h2_options = hcm.mutable_http2_protocol_options();
+          h2_options->mutable_max_concurrent_streams()->set_value(num_responses);
+          h2_options->mutable_initial_stream_window_size()->set_value(
+              buffer_params.downstream_h2_stream_window);
+          h2_options->mutable_initial_connection_window_size()->set_value(
+              buffer_params.downstream_h2_conn_window);
+        });
+
+    config_helper_.addConfigModifier(
+        [&](envoy::config::bootstrap::v3::Bootstrap& bootstrap) -> void {
+          ConfigHelper::HttpProtocolOptions protocol_options;
+          auto* upstream_h2_options =
+              protocol_options.mutable_explicit_http_config()->mutable_http2_protocol_options();
+          upstream_h2_options->mutable_max_concurrent_streams()->set_value(100);
+          upstream_h2_options->mutable_initial_stream_window_size()->set_value(
+              buffer_params.upstream_h2_stream_window);
+          upstream_h2_options->mutable_initial_connection_window_size()->set_value(
+              buffer_params.upstream_h2_conn_window);
+          for (auto& cluster_config : *bootstrap.mutable_static_resources()->mutable_clusters()) {
+            ConfigHelper::setProtocolOptions(cluster_config, protocol_options);
+          }
+        });
```

Comment:
The complicated H2 config above is not needed by any of the test cases you're using.  The buffer sizes involved are small; the default stream and connection windows work fine.  Consider removing the config modifiers above and associated parameters.